File: dotnet-advisor-core.Tests/Integration/IntegrationHelpers.cs
using System.Globalization;
using System.IO;
using StormGeo.AdvisorCore;
using StormGeo.AdvisorCore.Payloads;
using Xunit;

namespace StormGeo.AdvisorCore.Tests.Integration;

public sealed class IntegrationFactAttribute : FactAttribute
{
    public IntegrationFactAttribute()
    {
        IntegrationEnvironment.Load();
        if (string.IsNullOrWhiteSpace(Environment.GetEnvironmentVariable("ADVISOR_TOKEN")))
        {
            Skip = "Set ADVISOR_TOKEN before running the .NET integration tests.";
        }
    }
}

public sealed class IntegrationPayloads
{
    public required WeatherPayload WeatherPayload { get; init; }
    public required WeatherPayload WeatherChartPayload { get; init; }
    public required ClimatologyPayload ClimatologyPayload { get; init; }
    public required CurrentWeatherPayload CurrentWeatherPayload { get; init; }
    public required StationPayload StationPayload { get; init; }
    public required RadiusPayload RadiusPayload { get; init; }
    public required GeometryPayload GeometryPayload { get; init; }
    public required RadiusPayload LightningDetailsPayload { get; init; }
    public required LightningLitePayload LightningLitePayload { get; init; }
    public required StorageListPayload StorageListPayload { get; init; }
    public required PlanInfoPayload PlanInfoPayload { get; init; }
    public required RequestDetailsPayload RequestDetailsPayload { get; init; }
    public required StaticMapPayload StaticMapPayload { get; init; }
    public required TmsPayload TmsPayload { get; init; }
    public required PmtilesPayload PmtilesPayload { get; init; }
    public required string SchemaIdentifier { get; init; }
    public required Dictionary<string, SchameDefinitionField> SchemaDefinitionFields { get; init; }
    public required Dictionary<string, object> SchemaParameters { get; init; }
}

internal static class IntegrationEnvironment
{
    private static bool _loaded;

    public static void Load()
    {
        if (_loaded)
        {
            return;
        }

        _loaded = true;

        var envFile = FindIntegrationEnvFile(AppContext.BaseDirectory);
        if (envFile is null)
        {
            return;
        }

        foreach (var rawLine in File.ReadAllLines(envFile))
        {
            var line = rawLine.Trim();
            if (line.Length == 0 || line.StartsWith('#'))
            {
                continue;
            }

            var separatorIndex = line.IndexOf('=');
            if (separatorIndex < 0)
            {
                continue;
            }

            var name = line[..separatorIndex].Trim();
            if (string.IsNullOrWhiteSpace(name) || !string.IsNullOrWhiteSpace(Environment.GetEnvironmentVariable(name)))
            {
                continue;
            }

            var value = line[(separatorIndex + 1)..].Trim();
            if (value.Length >= 2)
            {
                var first = value[0];
                var last = value[^1];
                if ((first == '"' && last == '"') || (first == '\'' && last == '\''))
                {
                    value = value[1..^1];
                }
            }

            Environment.SetEnvironmentVariable(name, value);
        }
    }

    private static string? FindIntegrationEnvFile(string startDirectory)
    {
        var directory = new DirectoryInfo(startDirectory);
        while (directory is not null)
        {
            var candidate = Path.Combine(directory.FullName, ".env.integration.local");
            if (File.Exists(candidate))
            {
                return candidate;
            }

            directory = directory.Parent;
        }

        return null;
    }
}

public static class IntegrationHelpers
{
    public static AdvisorCore CreateSdk()
    {
        IntegrationEnvironment.Load();
        var token = Environment.GetEnvironmentVariable("ADVISOR_TOKEN");
        if (string.IsNullOrWhiteSpace(token))
        {
            throw new InvalidOperationException("Set ADVISOR_TOKEN before running the .NET integration tests.");
        }

        var sdk = new AdvisorCore(token, attempts: 1, delayInSeconds: 0);
        sdk.SetHeaderAccept("application/json");
        sdk.SetHeaderAcceptLanguage(Environment.GetEnvironmentVariable("ADVISOR_ACCEPT_LANGUAGE") ?? "en-US");
        return sdk;
    }

    public static IntegrationPayloads CreatePayloads()
    {
        IntegrationEnvironment.Load();
        var localeId = GetEnvInt("ADVISOR_LOCALE_ID", 3477);
        var stationId = GetRequiredEnv("ADVISOR_STATION_ID");
        var geometry = GetRequiredEnv("ADVISOR_GEOMETRY");
        var today = DateTime.Now;
        var observedDay = today.AddDays(-1);
        var observedPeriodEnd = observedDay;
        var observedPeriodStart = observedPeriodEnd.AddDays(-4);
        var forecastDay = today.AddDays(1);
        var forecastHourEnd = new DateTime(forecastDay.Year, forecastDay.Month, forecastDay.Day, 1, 0, 0, forecastDay.Kind);
        const string schemaIdentifier = "schemaIdentifier";

        return new IntegrationPayloads
        {
            WeatherPayload = new WeatherPayload
            {
                LocaleId = localeId,
                Variables = ["temperature"],
            },
            WeatherChartPayload = new WeatherPayload
            {
                LocaleId = localeId,
                Variables = ["temperature", "precipitation"],
            },
            ClimatologyPayload = new ClimatologyPayload
            {
                LocaleId = localeId,
                Variables = ["temperature"],
            },
            CurrentWeatherPayload = new CurrentWeatherPayload
            {
                LocaleId = localeId,
            },
            StationPayload = new StationPayload
            {
                StationId = stationId,
            },
            RadiusPayload = new RadiusPayload
            {
                LocaleId = localeId,
                Radius = 10000,
            },
            GeometryPayload = new GeometryPayload
            {
                Geometry = geometry,
                StartDate = StartOfDay(observedDay),
                EndDate = EndOfDay(observedDay),
                Radius = 10000,
            },
            LightningDetailsPayload = new RadiusPayload
            {
                Latitude = "-22.9",
                Longitude = "-43.2",
                StartDate = StartOfDay(observedDay),
                EndDate = EndOfDay(observedDay),
                Radius = 10000,
            },
            LightningLitePayload = new LightningLitePayload
            {
                Geometry = geometry,
                StartDate = StartOfDay(observedPeriodStart),
                EndDate = EndOfDay(observedPeriodEnd),
                Radius = 10000,
                Page = 1,
                PageSize = 50,
            },
            StorageListPayload = new StorageListPayload(1, 10),
            PlanInfoPayload = new PlanInfoPayload
            {
                Timezone = -3,
            },
            RequestDetailsPayload = new RequestDetailsPayload(1, 3),
            StaticMapPayload = new StaticMapPayload
            {
                Type = "periods",
                Category = "observed",
                Variable = "temperature",
                Aggregation = "max",
                StartDate = StartOfDay(observedPeriodStart),
                EndDate = EndOfDay(observedPeriodEnd),
                Dpi = 50,
                Title = true,
                TitleVariable = "Static Map",
            },
            TmsPayload = new TmsPayload
            {
                Server = 'a',
                Mode = "forecast",
                Variable = "precipitation",
                Aggregation = "sum",
                X = 5,
                Y = 8,
                Z = 4,
                Istep = StartOfDay(forecastDay),
                Fstep = EndOfDay(forecastDay),
            },
            PmtilesPayload = new PmtilesPayload
            {
                Mode = "forecast",
                Model = "ct2w15_as",
                Variable = "precipitation",
                Aggregation = "sum",
                Istep = StartOfDay(forecastDay),
                Fstep = FormatDateTime(forecastHourEnd),
                MaxZoom = 4,
            },
            SchemaIdentifier = schemaIdentifier,
            SchemaDefinitionFields = new Dictionary<string, SchameDefinitionField>
            {
                ["arbitraryField1"] = new SchameDefinitionField
                {
                    Type = "boolean",
                    Required = true,
                    Length = 125,
                },
                ["arbitraryField2"] = new SchameDefinitionField
                {
                    Type = "number",
                    Required = true,
                },
                ["arbitraryField3"] = new SchameDefinitionField
                {
                    Type = "string",
                    Required = false,
                },
            },
            SchemaParameters = new Dictionary<string, object>
            {
                ["arbitraryField1"] = true,
                ["arbitraryField2"] = 15,
            },
        };
    }

    public static StorageDownloadPayload ResolveStorageDownloadPayload()
    {
        IntegrationEnvironment.Load();
        return new StorageDownloadPayload
        {
            FileName = GetRequiredEnv("ADVISOR_STORAGE_FILE_NAME"),
            AccessKey = GetRequiredEnv("ADVISOR_STORAGE_ACCESS_KEY"),
        };
    }

    public static void AssertJsonSuccess(AdvisorResponse<string> response)
    {
        Assert.Null(response.Error);
        Assert.False(string.IsNullOrWhiteSpace(response.Data));
    }

    public static async Task AssertStreamSuccessAsync(AdvisorResponse<Stream> response)
    {
        Assert.Null(response.Error);
        Assert.NotNull(response.Data);

        await using var stream = response.Data!;
        using var memory = new MemoryStream();
        await stream.CopyToAsync(memory);
        Assert.NotEmpty(memory.ToArray());
    }

    private static int GetEnvInt(string name, int defaultValue)
    {
        var value = Environment.GetEnvironmentVariable(name);
        return int.TryParse(value, out var parsed) ? parsed : defaultValue;
    }

    private static string GetRequiredEnv(string name)
    {
        var value = Environment.GetEnvironmentVariable(name);
        if (!string.IsNullOrWhiteSpace(value))
        {
            return value;
        }

        throw new InvalidOperationException(
            $"Set {name} or add it to .env.integration.local before running the .NET integration tests.");
    }

    private static string StartOfDay(DateTime value)
    {
        return FormatDateTime(new DateTime(value.Year, value.Month, value.Day, 0, 0, 0, value.Kind));
    }

    private static string EndOfDay(DateTime value)
    {
        return FormatDateTime(new DateTime(value.Year, value.Month, value.Day, 23, 59, 59, value.Kind));
    }

    private static string FormatDateTime(DateTime value)
    {
        return value.ToString("yyyy-MM-dd HH:mm:ss", CultureInfo.InvariantCulture);
    }
}

File: node-advisor-core/tests/integration.helpers.ts
import assert from "node:assert/strict"
import { existsSync, readFileSync } from "node:fs"
import path from "node:path"
import { Readable } from "node:stream"

import { AdvisorCore } from "../src/AdvisorCore"
import {
  ApiFileResponse,
  ApiResponse,
  ApiStreamResponse,
  ClimatologyPayload,
  CurrentWeatherPayload,
  GeometryPayload,
  PlanInfoPayload,
  PlanLocalePayload,
  PmtilesPayload,
  RadiusPayload,
  RequestDetailsPayload,
  StationPayload,
  StationsLastDataPayload,
  StaticMapPayload,
  StorageDownloadPayload,
  StorageListPayload,
  TmsPayload,
  WeatherPayload,
} from "../src/payloads"

type JsonLike = Record<string, unknown>

export interface IntegrationPayloads {
  weatherPayload: WeatherPayload
  weatherChartPayload: WeatherPayload
  climatologyPayload: ClimatologyPayload
  currentWeatherPayload: CurrentWeatherPayload
  stationPayload: StationPayload
  stationsLastDataPayload: StationsLastDataPayload
  radiusPayload: RadiusPayload
  geometryPayload: GeometryPayload
  lightningDetailsPayload: RadiusPayload
  lightningLitePayload: GeometryPayload & { page: number; pageSize: number }
  storageListPayload: StorageListPayload
  planInfoPayload: PlanInfoPayload
  planLocalePayload: PlanLocalePayload
  requestDetailsPayload: RequestDetailsPayload
  staticMapPayload: StaticMapPayload
  tmsPayload: TmsPayload
  pmtilesPayload: PmtilesPayload
  schemaDefinitionPayload: JsonLike
}

function getEnvInt(name: string, defaultValue: number): number {
  const value = process.env[name]
  return value ? Number.parseInt(value, 10) : defaultValue
}

function findIntegrationEnvFile(startDir: string): string | undefined {
  let currentDir = startDir

  while (true) {
    const candidate = path.join(currentDir, ".env.integration.local")
    if (existsSync(candidate)) {
      return candidate
    }

    const parentDir = path.dirname(currentDir)
    if (parentDir === currentDir) {
      return undefined
    }

    currentDir = parentDir
  }
}

function loadIntegrationEnv(): void {
  const envFile = findIntegrationEnvFile(__dirname) ?? findIntegrationEnvFile(process.cwd())
  if (!envFile) {
    return
  }

  for (const rawLine of readFileSync(envFile, "utf8").split(/\r?\n/u)) {
    const line = rawLine.trim()
    if (!line || line.startsWith("#")) {
      continue
    }

    const separatorIndex = line.indexOf("=")
    if (separatorIndex === -1) {
      continue
    }

    const name = line.slice(0, separatorIndex).trim()
    if (!name || process.env[name]) {
      continue
    }

    let value = line.slice(separatorIndex + 1).trim()
    if (
      value.length >= 2
      && ((value.startsWith("\"") && value.endsWith("\"")) || (value.startsWith("'") && value.endsWith("'")))
    ) {
      value = value.slice(1, -1)
    }

    process.env[name] = value
  }
}

function requireEnv(name: string): string {
  const value = process.env[name]
  if (!value) {
    throw new Error(
      `Set ${name} or add it to .env.integration.local before running the Node integration tests.`,
    )
  }
  return value
}

loadIntegrationEnv()

function pad(value: number): string {
  return String(value).padStart(2, "0")
}

function formatDateTime(value: Date): string {
  return [
    value.getFullYear(),
    pad(value.getMonth() + 1),
    pad(value.getDate()),
  ].join("-") + " " + [
    pad(value.getHours()),
    pad(value.getMinutes()),
    pad(value.getSeconds()),
  ].join(":")
}

function startOfDay(value: Date): string {
  return formatDateTime(new Date(
    value.getFullYear(),
    value.getMonth(),
    value.getDate(),
    0,
    0,
    0,
  ))
}

function endOfDay(value: Date): string {
  return formatDateTime(new Date(
    value.getFullYear(),
    value.getMonth(),
    value.getDate(),
    23,
    59,
    59,
  ))
}

function shiftDays(value: Date, amount: number): Date {
  const shifted = new Date(value)
  shifted.setDate(shifted.getDate() + amount)
  return shifted
}

export function createAdvisor(): AdvisorCore {
  const advisor = new AdvisorCore({
    token: requireEnv("ADVISOR_TOKEN"),
    retries: 1,
    delay: 0,
  })

  advisor.setHeaderAccept("application/json")
  advisor.setHeaderAcceptLanguage(process.env.ADVISOR_ACCEPT_LANGUAGE ?? "en-US")

  return advisor
}

export function createPayloads(): IntegrationPayloads {
  const localeId = getEnvInt("ADVISOR_LOCALE_ID", 3477)
  const planLocaleId = getEnvInt("ADVISOR_PLAN_LOCALE_ID", 5959)
  const stationId = requireEnv("ADVISOR_STATION_ID")
  const geometry = requireEnv("ADVISOR_GEOMETRY")
  const today = new Date()
  const observedDay = shiftDays(today, -1)
  const observedPeriodEnd = observedDay
  const observedPeriodStart = shiftDays(observedPeriodEnd, -4)
  const forecastDay = shiftDays(today, 1)
  const forecastHourEnd = new Date(
    forecastDay.getFullYear(),
    forecastDay.getMonth(),
    forecastDay.getDate(),
    1,
    0,
    0,
  )
  const schemaIdentifier = "schemaIdentifier"

  return {
    weatherPayload: {
      localeId,
      variables: ["temperature"],
    },
    weatherChartPayload: {
      localeId,
      variables: ["temperature", "precipitation"],
    },
    climatologyPayload: {
      localeId,
      variables: ["temperature"],
    },
    currentWeatherPayload: {
      localeId,
    },
    stationPayload: {
      stationId,
    },
    stationsLastDataPayload: {
      stationIds: [stationId],
      variables: ["temperature"],
    },
    radiusPayload: {
      localeId,
      radius: 10000,
    },
    geometryPayload: {
      geometry,
      startDate: startOfDay(observedDay),
      endDate: endOfDay(observedDay),
      radius: 10000,
    },
    lightningDetailsPayload: {
      latitude: -22.9,
      longitude: -43.2,
      startDate: startOfDay(observedDay),
      endDate: endOfDay(observedDay),
      radius: 10000,
    },
    lightningLitePayload: {
      geometry,
      startDate: startOfDay(observedPeriodStart),
      endDate: endOfDay(observedPeriodEnd),
      radius: 10000,
      page: 1,
      pageSize: 50,
    },
    storageListPayload: {
      page: 1,
      pageSize: 10,
    },
    planInfoPayload: {
      timezone: -3,
    },
    planLocalePayload: {
      localeId: planLocaleId,
    },
    requestDetailsPayload: {
      page: 1,
      pageSize: 3,
    },
    staticMapPayload: {
      type: "periods",
      category: "observed",
      variable: "temperature",
      aggregation: "max",
      startDate: startOfDay(observedPeriodStart),
      endDate: endOfDay(observedPeriodEnd),
      dpi: 50,
      title: true,
      titlevariable: "Static Map",
    },
    tmsPayload: {
      istep: startOfDay(forecastDay),
      fstep: endOfDay(forecastDay),
      server: "a",
      mode: "forecast",
      variable: "precipitation",
      aggregation: "sum",
      x: "5",
      y: "8",
      z: "4",
    },
    pmtilesPayload: {
      mode: "forecast",
      model: "ct2w15_as",
      variable: "precipitation",
      aggregation: "sum",
      istep: startOfDay(forecastDay),
      fstep: formatDateTime(forecastHourEnd),
      maxZoom: 4,
    },
    schemaDefinitionPayload: {
      identifier: schemaIdentifier,
      arbitraryField1: {
        type: "boolean",
        required: true,
        length: 125,
      },
      arbitraryField2: {
        type: "number",
        required: true,
      },
      arbitraryField3: {
        type: "string",
        required: false,
      },
    }
  }
}

export async function resolveStorageDownloadPayload(
  _advisor: AdvisorCore,
  _storageListPayload: StorageListPayload,
): Promise<StorageDownloadPayload> {
  return {
    fileName: requireEnv("ADVISOR_STORAGE_FILE_NAME"),
    accessKey: requireEnv("ADVISOR_STORAGE_ACCESS_KEY"),
  }
}

function formatError(error: unknown): string {
  if (!error) {
    return "Unknown error"
  }

  if (typeof error === "string") {
    return error
  }

  if (error instanceof Error) {
    return error.message
  }

  return JSON.stringify(error)
}

export function assertJsonSuccess(response: ApiResponse): void {
  assert.equal(response.error, null, formatError(response.error))
  assert.notEqual(response.data, null)
}

export function assertBufferSuccess(response: ApiFileResponse): void {
  assert.equal(response.error, null, formatError(response.error))
  assert.ok(Buffer.isBuffer(response.data))
  assert.ok(response.data.length > 0)
}

export async function assertStreamSuccess(response: ApiStreamResponse): Promise<void> {
  assert.equal(response.error, null, formatError(response.error))
  assert.ok(response.data)

  const chunks: Buffer[] = []
  for await (const chunk of response.data as Readable) {
    chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk))
  }

  assert.ok(chunks.length > 0)
  assert.ok(Buffer.concat(chunks).length > 0)
}
